File: C++/reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        vector<ListNode*> tmp;
        
        for (ListNode* n = head; n; n = n->next) {
            tmp.push_back(n);
        }
        
        int i = 0, j = tmp.size() - 1;
        
        while (true) {
            if (i < j) {
                tmp[i]->next = tmp[j];
            } else {
                tmp[i]->next = nullptr;
                break;
            }
            i++;
            if (i < j) {
                tmp[j]->next = tmp[i];
            } else {
                tmp[j]->next = nullptr;
                break;
            }
            j--;
        }
    }
};

File: Kotlin/reorder-list.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun reorderList(head: ListNode?): Unit {
        if (head == null) {
            return
        }
        var elements = mutableListOf<ListNode>()        
        var node = head
        while (node != null) {
            val next = node.next
            node.next = null
            elements.add(node)
            node = next
        }
        var i = 0
        var j = elements.size - 1
        while (i < j) {
            elements[i++].next = elements[j]
            if (i < j)
                elements[j--].next = elements[i]
        }
    }
}